### ISSUE TITLE
Deduplicate in-flight fetchAuthSession token refresh requests

### ISSUE THREAD

**Issue by OrmEmbaar** [aws-amplify/amplify-js#13499]
### Is this related to a new or existing API?

Authentication, PubSub

### Is this related to another service?

Cognito

### Describe the feature you'd like to request

We are using the Next Pages Router and are in the process of upgrading from aws-amplify v4 to v6.

On initial page load, we fetch data from our server using a signed Authorization header from the SignatureV4 class and setup signed MQTT connections using the PubSub library. For every request we make to our server, we're manually calling `fetchAuthSession` to get the credentials for the signed Authorization header, and for every MQTT connection we establish, the PubSub library calls `fetchAuthSession` internally. As this all happens on initial page load, we're firing multiple concurrent `fetchAuthSession` requests before the singleton instance has received any credentials from the Cognito server.

This pattern was working fine with `Auth.currentCredentials` on the v4 library. However, when using `fetchAuthSession` on v6, we hit the Cognito server every single time and are therefore being rate-limited.

We therefore would like to see the `fetchAuthSession` singleton de-duplicate concurrent in-flight requests to the Cognito server.  

### Describe the solution you'd like

The aws-amplify library should await any in-flight requests to the Cognito server instead of making duplicate concurrent requests. 

### Describe alternatives you've considered

Calling `fetchAuthSession` once on application load then passing the returned credentials around. However, that is problematic for a few reasons:

1. We would also need to build a mechanism to refresh stale tokens.
2. The PubSub library handles credentials itself, so we can't pass it tokens set on state. 

### Is this something that you'd be interested in working on?

- [ ] 👋 I may be able to implement this feature request
- [ ] ⚠️ This feature might incur a breaking change

**Comment by cwomack** [comment]
Hello, @OrmEmbaar and thanks for creating this feature request. I'll review it with the team internally and follow up with any additional questions we have.

**Comment by mkolbusz** [comment]
@cwomack I think I have encountered the same problem while using Tanstack Query with NextJs. After session tokens have expired and Tanstack Query is trying to refetch the data the server multiplies the cookies and tokens as presented below: 
![Selection_083](https://github.com/aws-amplify/amplify-js/assets/4622240/c5d40304-f69a-4993-be85-0f7e09c13564)

Minimum repository with reproduction: https://github.com/mkolbusz/nextjs-amplify-v6-issues

Steps to reproduce:
1. Login.
2. Go to the other tab in the browser.
3. Wait for the session to expire.
4. Enter the tab of the application (refetching data and refreshing the session at the same time).
5. The auth cookies are multiplicated.

This problem does not allow to sign out properly.

**Comment by OrmEmbaar** [comment]
It looks like the library already has tools needed to make this work. This `deDupeAsyncFunction` function is used to wrap the `refreshToken` method.

https://github.com/aws-amplify/amplify-js/blob/de6b06db80b02687e597aceaa2c33f65171a6389/packages/core/src/utils/deDupeAsyncFunction.ts#L10

But that is only called when refreshing tokens, not when fetching credentials. To de-duplicate initial credential fetch, the `getId` function would also need to be wrapped by `deDupeAsyncFunction`:

https://github.com/aws-amplify/amplify-js/blob/de6b06db80b02687e597aceaa2c33f65171a6389/packages/core/src/awsClients/cognitoIdentity/getId.ts#L56

It seems like an easy change. Can we make this happen?

**Comment by mkolbusz** [comment]
@jimblanc issue created: https://github.com/aws-amplify/amplify-js/issues/13509

**Comment by OrmEmbaar** [comment]
A workaround 

```typescript
import { AuthSession, fetchAuthSession } from 'aws-amplify/auth';

/**
 * This function creates a singleton that fetches the current session. This is necessary
 * because the Amplify fetchAuthSession method does not de-duplicate credential requests to
 * the Cognito server. That means that duplicate requests will be made to the server if
 * multiple components call fetchAuthSession concurrently before the internal credential
 * cache has been populated. This singleton ensures that only one request is made at a time.
 *
 * A feature request has been made to the Amplify team to add de-duplication
 * to the fetchAuthSession method. If that feature is added, this layer of
 * abstraction can be removed.
 *
 * @see https://github.com/aws-amplify/amplify-js/issues/13499
 */
function createFetchAuthSessionDedupeSingleton() {
  let pendingRequest: Promise<AuthSession> | null = null;

  return () => {
    if (!pendingRequest) {
      pendingRequest = new Promise(async (resolve, reject) => {
        try {
          const response = await fetchAuthSession();
          resolve(response);
        } catch (error) {
          reject(error);
        } finally {
          pendingRequest = null;
        }
      });
    }

    return pendingRequest;
  };
}

export const fetchAuthSessionDedupe = createFetchAuthSessionDedupeSingleton();
```

**Comment by didemkkaslan** [comment]
@OrmEmbaar  how about the serverside fetchAuthSession? 
`import { fetchAuthSession } from 'aws-amplify/auth/server'`
 Did you do the same deduplication for it too?

**Comment by OrmEmbaar** [comment]
@didemkkaslan No, I was concerned that a singleton on the server would leak credentials across requests. Also, I don't have any concurrent authentication requests happening on the server during page load, so it wasn't necessary.

**Comment by didemkkaslan** [comment]
Got it thanks @OrmEmbaar, do you think this will also fix the tokens getting multiplied after session tokens expired problem mkolbusz mentions. 

https://github.com/aws-amplify/amplify-js/issues/13499#issuecomment-2174403980

**Comment by OrmEmbaar** [comment]
That issue should have been resolved in a previous update https://github.com/aws-amplify/amplify-js/issues/13509#issuecomment-2244047261. Are you on the latest version?

**Comment by didemkkaslan** [comment]
Yes I'am but this still is the case for me. 
```
  "aws-amplify": "^6.6.4",
 "@aws-amplify/adapter-nextjs": "^1.2.21",
```


<img width="1009" alt="Screenshot 2024-10-22 at 16 07 31" src="https://github.com/user-attachments/assets/55bbfc11-4f03-4284-aa6b-a0f7dd34d365">

Maybe its because of my custom cookie storage can't really find the problem btw env variable is prod so its {} for the cookie options. Tokens get multiplied for the same path /


```
const cookieOptions: OptionsType =
  process.env.NEXT_PUBLIC_ENV === 'msteams'
    ? {
        domain: 'tab.app.spiky.ai' as string,
        sameSite: 'none' as 'lax' | 'strict' | 'none',
        secure: true,
      }
    : {};

const keyValueStorage = createKeyValueStorageFromCookieStorageAdapter({
  get(name) {
    const value = getCookie(name, cookieOptions);
    return { name, value };
  },
  getAll() {
    const cookies = getCookies(cookieOptions);
    return Object.keys(cookies).map((name) => ({ name, value: cookies[name] }));
  },
  set(name, value) {
    setCookie(name, value, cookieOptions);
  },
  delete(name) {
    deleteCookie(name, cookieOptions);
  },
});
```

**Comment by OrmEmbaar** [comment]
@didemkkaslan From the cookie names it looks like you have two sets of cookies from two different users, which is different to the issue reported by mkolbusz above. Is it causing issues in your application? Perhaps your storage adapter is not clearing them properly.

**Comment by HuiSF** [comment]
Hi @didemkkaslan, echo to what @OrmEmbaar it seems you had an end user signed in before and didn't sign out before signing in a different user. It's unrelated to the issue described in the OP. And it's an different issue from https://github.com/aws-amplify/amplify-js/issues/13509 that has been fixed.
